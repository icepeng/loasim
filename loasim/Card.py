from typing import List, Tuple
from loasim.CharacterModifier import CharacterModifier


class Card:
    def __init__(self, name: str, mdf: CharacterModifier) -> None:
        self.name = name
        self.mdf = mdf

    def get_modifier(self):
        return self.mdf


card_list = [
    Card("남겨진 바람의 절벽 (12)", CharacterModifier(crit=7)),
    Card("세상을 구하는 빛 (18)", CharacterModifier(pdamage_indep=7)),
    Card("세상을 구하는 빛 (30)", CharacterModifier(pdamage_indep=15)),
]

card_dict = {card.name: card for card in card_list}


class CardManager:
    def __init__(self, *equipped_list: List[str]) -> None:
        self.equipped_list = equipped_list

    def get_static_modifier(self):
        mdf = CharacterModifier()
        for name in self.equipped_list:
            card = card_dict.get(name)
            if card is None:
                raise Exception(f"{name} is not an card")
            mdf = mdf + card.get_modifier()
        return mdf
