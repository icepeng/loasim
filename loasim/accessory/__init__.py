from loasim.accessory.builtin import lostark_accessory_repository
from loasim.accessory.base import AccessoryState
