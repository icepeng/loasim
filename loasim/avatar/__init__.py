from loasim.avatar.base import AvatarState
from loasim.avatar.builtin import lostark_avatar_repository
