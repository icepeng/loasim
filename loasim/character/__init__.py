from loasim.character.base import AccessoryState, AvatarState, Character, GearState
