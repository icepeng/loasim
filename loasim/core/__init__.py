from loasim.core.buff import (
    BuffManager,
    OnoffBuff,
    SkillBuff,
    StackBuff,
    StatBuff,
    StaticBuff,
)
from loasim.core.enemy import Enemy
from loasim.core.skill import Skill, SkillSpecification, Tripod
from loasim.core.stat import InternalStat, Stat
