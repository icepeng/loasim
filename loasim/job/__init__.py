from loasim.job.artist import lostark_job_artist
from loasim.job.base import Job, SkillSpecification, SkillState, Tripod
from loasim.job.blade import lostark_job_blade
from loasim.job.hawkeye import lostark_job_hawkeye
from loasim.job.scouter import lostark_job_scouter

job_dict = {
    "artist": lostark_job_artist,
    "blade": lostark_job_blade,
    "hawkeye": lostark_job_hawkeye,
    "scouter": lostark_job_scouter,
}


def get_job(job_name: str) -> Job:
    job = job_dict[job_name]
    if job is None:
        raise ValueError(f"{job_name} is not supported")
    return job
