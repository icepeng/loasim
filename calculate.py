from collections import defaultdict
from typing import List
from loasim.core import Skill


def calculate(dealcycle: List[Skill], cycle_time, enemy):
    total_damage = 0
    damage_dict = defaultdict(float)
    for sk in dealcycle:
        damage = sk.get_damage(enemy)
        total_damage += damage
        damage_dict[sk.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / cycle_time:,.0f}")
