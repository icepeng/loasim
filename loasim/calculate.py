from loguru import logger

from collections import defaultdict
from typing import Any, Callable, List, Dict, Literal, Tuple

from pydantic import BaseModel

from loasim.core import Skill, Enemy, BuffManager, InternalStat
from loasim.core.buff import Buff
from loasim.core.stat import Stat
from loasim.setitem import lostark_setitem_repository
from loasim.card import lostark_default_card_repository
from loasim.engraving import lostark_engraving_repository


class DealCycle(BaseModel):
    buff_list: List[Buff]
    skill_list: List[Tuple[Skill, Dict[str, Any], Literal["back", "head"] | None]]
    cycle_time: float


def calculate(
    internal_stat: InternalStat,
    weapon_pdamage: float,
    setitem_state: List[Tuple[str, int, int]],
    engraving_state: List[Tuple[str, int]],
    card_state: List[str],
    generator: Callable[[InternalStat], DealCycle],
    enemy: Enemy,
):
    deal_cycle = generator(internal_stat)
    base_stat = internal_stat.get_stat()
    weapon_stat = Stat(pdamage=weapon_pdamage)  # 무기 품질
    basis_stat = (
        base_stat
        + weapon_stat
        + lostark_setitem_repository.get_stat(setitem_state)
        + lostark_default_card_repository.get_stat(card_state)
    )

    logger.info(base_stat)
    logger.info(basis_stat)

    engraving_buffs = lostark_engraving_repository.get_buffs(engraving_state)
    buff_manager = BuffManager(engraving_buffs + deal_cycle.buff_list)

    total_damage = 0.0
    damage_dict: defaultdict[str, float] = defaultdict(float)
    for skill, buff_state, backhead in deal_cycle.skill_list:
        buff_stat = buff_manager.get_stat(buff_state, skill)
        damage = skill.get_damage(enemy, backhead, basis_stat + buff_stat)
        total_damage += damage
        damage_dict[skill.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(deal_cycle.cycle_time)
    print(f"dps: {total_damage / deal_cycle.cycle_time:,.0f}")
