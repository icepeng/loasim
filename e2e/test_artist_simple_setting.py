from loasim.calculate import DealCycle, calculate
from loasim.character import AccessoryState, AvatarState, Character, GearState
from loasim.core import BuffState, Enemy
from loasim.job import SkillState


def calcuate_artist():
    state = {
        "저무는 달": BuffState(onoff=False),
        "내면의 각성": BuffState(onoff=True),
        "나만의 권능": BuffState(onoff=True),
        "나만의 우물": BuffState(onoff=True),
    }

    deal_cycle: DealCycle = [
        ("필법 : 한획긋기", state, None),
        ("묵법 : 두루미나래", state, None),
        ("묵법 : 범가르기", state, None),
        ("묵법 : 달그리기", state, None),
        ("묵법 : 먹오름", state, None),
    ]

    character = Character(
        combat_level=59,
        expedition_stat=637,
        crit=537,
        special=56,
        swift=1691,
        gear_state=GearState(
            head=(1340, "relic", 19),
            shoulder=(1340, "relic", 19),
            top=(1340, "relic", 18),
            bottom=(1340, "relic", 18),
            glove=(1340, "relic", 19),
            weapon=(1340, "relic", 19),
        ),
        accessory_state=AccessoryState(
            necklace="relic",
            ear1="relic",
            ear2="relic",
            ring1="relic",
            ring2="relic",
        ),
        avatar_state=AvatarState(
            head="epic",
            top="epic",
            bottom="epic",
            weapon="epic",
        ),
    )

    return calculate(
        job_name="artist",
        internal_stat=character.get_internal_stat(),
        weapon_pdamage=21.86,
        setitem_state=[
            ("악몽", 2, 0),
            ("지배", 4, 0),
        ],
        engraving_state=[
            ("원한", 3),
            ("예리한 둔기", 3),
            ("안정된 상태", 3),
            ("돌격대장", 3),
            ("회귀", 3),
        ],
        card_state=["남겨진 바람의 절벽 (12)"],
        skill_state={
            "묵법 : 해그리기": SkillState(
                level=10,
                tripod={
                    "나만의 권능": 5,
                },
            ),
            "묵법 : 해우물": SkillState(
                level=10,
                tripod={
                    "나만의 우물": 1,
                },
            ),
            "필법 : 한획긋기": SkillState(
                level=11,
                gem=7,
                tripod={
                    "거대한 붓": 5,
                    "강화된 일격": 5,
                },
            ),
            "묵법 : 두루미나래": SkillState(
                level=11,
                gem=5,
                tripod={
                    "치명적인 일격": 5,
                    "학익진": 5,
                },
            ),
            "묵법 : 범가르기": SkillState(
                level=11,
                gem=6,
                tripod={
                    "궤뚫는 일격": 5,
                    "강화된 일격": 5,
                },
            ),
            "묵법 : 달그리기": SkillState(
                level=11,
                gem=7,
                tripod={
                    "별 그리기": 4,
                    "붉은 달": 4,
                },
            ),
            "묵법 : 옹달샘": SkillState(
                level=11,
                gem=7,
                tripod={
                    "급소타격": 4,
                    "강화된 일격": 4,
                    "잉어 봉인해제!": 4,
                },
            ),
            "묵법 : 먹오름": SkillState(
                level=11,
                gem=5,
                tripod={
                    "치명적인 일격": 1,
                    "먹물점정": 1,
                },
            ),
        },
        enemy=Enemy(armor=6000, reduction=23),
        deal_cycle=deal_cycle,
        cycle_time=12.127,
    )


def test_artist_simple_setting(snapshot: float):
    assert calcuate_artist() == snapshot


if __name__ == "__main__":
    calcuate_artist()
