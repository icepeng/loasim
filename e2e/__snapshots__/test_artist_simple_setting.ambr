# name: test_artist_simple_setting
  2313644.5603710483
---
